### ISSUE TITLE
CompilationEnd action is not triggered in-IDE (VSMac)

### ISSUE THREAD

**Issue by Therzok** [dotnet/roslyn#30764]
**Version Used**: 2.10.0-beta2-63315-04

**Steps to Reproduce**:

1. Take [ReviewUnusedParameters](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/ReviewUnusedParameters.cs#L50) analyzer.
2. Observe that no action from the CompilationStart is triggered (most importantly, CompilationEnd)

**Expected Behavior**:
CompilationEnd action should be triggered in all cases

**Actual Behavior**:
When full solution analysis is off, it is never triggered.

From my analysis, what happens is that the compilation only has one tree added when FSA is off, and that is the filtered tree of the current document. See the following stacktraces.
https://gist.github.com/Therzok/76bd78cb68e25de24bc9d893fb18cd47

Shouldn't syntax trees still not be filtered to current document when FSA is off? An analyzer could be running in the current document, but still need access to the full solution state.

Not sure how to investigate this further, but I hope this is enough.

**Comment by Therzok** [comment]
Is it because each analyzer makes a fork of the compilation at the point where no syntax tree is computed?

http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Compilation/CSharpCompilation.cs,1952

That means that every analyzer will start with a compilation that does not have the map created, thus will never have all the syntax trees filled.

**Comment by Therzok** [comment]
In the end, it looks like every `CompilationWithAnalyzers` will have a [new set of syntax trees to complete](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,125) which will never be filled.

**Comment by CyrusNajmabadi** [comment]
@mavasani @heejaechang .  I believe it is "by limitation", namely that it's too expensive otherwise.

**Comment by Therzok** [comment]
Doesn't the backgroundcompiler already do full compilation of every project that the document's project depends on?

Why is it a limitation in that case?

**Comment by CyrusNajmabadi** [comment]
> Doesn't the backgroundcompiler already do full compilation of every project that the document's project depends on?

No.  First, if you are C#->C# or VB->VB, it just does a direct compilation reference.  If you are C#->VB or VB->C#, it will build just the *metadata* for that compiation (i.e. the declaration signatures).

**Comment by CyrusNajmabadi** [comment]
And, also: 

> When full solution analysis is off, it is never triggered.

When you have fsa off... that's sorta the point of what's going on.  It's not analyzing things 'fully', it's only trying to analyze open files.  So running the CompilationEnd action is going to be highly problematic in many cases, because the rest of your analyzers will have only analyzed part of the compilation.  And, if you had them analyze everything, then that would somewhat defeat the purpose of FSA.

**Comment by CyrusNajmabadi** [comment]
Dupe of https://github.com/dotnet/roslyn/issues/11750?

**Comment by heejaechang** [comment]
@Therzok if everyone agrees that analyzer (diagnostics) with not so good code fix experience is fine. then we can enable it for open files.

otherwise, we need to fix up code fix experience before we can turn it on.

...

I won't get into the not so good code fix experience since whatever people come up so far, there are other people who don't like it. if we want to see what is the not so good code fix experience, just turn full solution analysis on, and try to use code fix for compilation end analyzer in a reasonable sized solution.

basically, whatever we come up with, it doesn't change the fact that calculating compilation end analyzer could be quite expensive, so we can't provide experience non compilation end analyzer code fix can provide now (the speed, lightness and etc), and any attempt to workaround it (like making code fix list async and etc) got shot down. so. we have no place to go I guess until desktop computer CPU gets like 10x times faster or something.

moving it to server farm or cloud to get unlimited CPU time is also got shutdown due to complexity it will bring in, and operation cost.

so, literally, can't move anywhere.

probably best way is someone just prototype it outside of roslyn so it can't get shutdown by other people.

**Comment by Therzok** [comment]
Understood. I was replacing one analyzer we had because it was really expensive: https://github.com/mono/RefactoringEssentials/blob/monodevelop2/RefactoringEssentials/CSharp/Diagnostics/Synced/RedundanciesInDeclaration/UnusedParameterAnalyzer.cs

and come to think of it, with the above we'd just revert to the old behaviour, just a bit lazier.
